Change how SPO votes are counted

Lucsanszky opened the pull request:
# Description
Resolves #4645 
<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

Lucsanszky committed:
Change how SPO votes are counted

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -201,10 +207,16 @@ spoAcceptedRatio
+                  | otherwise -> (yes, abstain)
+                Just DRepAlwaysAbstain -> (yes, abstain + stake)
+                -- TODO: figure out if other cases need special care
+                _ -> (yes, abstain)
```
We need to make sure that:
* we do not change the behavior of `HardForkInitiation`
* we preserve the bootstrap behavior
* we have access to protocol version, which we probably can also stick into `RatifyEnv`
```suggestion
              Nothing
                | HardForkInitiation {} <- pProcGovAction -> (yes, abstain)
                | bootstrapPhase pv -> (yes, abstain + stake)
                | otherwise -> case res of
                  Just DRepAlwaysNoConfidence
                    | NoConfidence {} <- pProcGovAction -> (yes + stake, abstain)
                  Just DRepAlwaysAbstain -> (yes, abstain + stake)
                  _ -> (yes, abstain)
```

lehins reviewed (commented):
This is looking good

Lucsanszky committed:
TEMP - Vote 'Abstain' w/ SPOs in conformance tests

Blocked by: https://github.com/IntersectMBO/formal-ledger-specifications/issues/578

Lucsanszky commented:
~I had to disable the ratify conformance test until https://github.com/IntersectMBO/formal-ledger-specifications/issues/578 is resolved, since now the spec and the implementation differs with regards to the SPO vote counts.~ I managed to tweak the SPO votes for conformance tests to make the tests pass until the spec is up-to-date. I'll probably add some more comments why I needed to adjust some of the stake amounts in pools for the tests (`electBasicCommittee` registers a stake pool and with the vote count changes we didn't always reach the threshold post-bootstrap). Other than that, this is ready for review and I'll add some tests in a separate PR.

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -187,24 +193,36 @@ spoAccepted re rs gas =
                 | HardForkInitiation {} <- pProcGovAction -> (yes, abstain)
-                | otherwise -> (yes, abstain + stake)
+                | bootstrapPhase pv -> (yes, abstain + stake)
+                | otherwise -> case drep of
```
Would you mind either adding a comment here, explaining the idea - or maybe better yet, updating the haddock above to reflect this case.

teodanciu commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs:
```diff
@@ -187,24 +193,36 @@ spoAccepted re rs gas =
         let CompactCoin stake = individualTotalPoolStake distr
             vote = Map.lookup poolId gasStakePoolVotes
+            drep =
+              Map.lookup poolId rePoolParams >>= \d -> Map.lookup ((raCredential . ppRewardAccount) d) reDelegatees
```
I wonder if  pattern matching would be more efficient, since this will be performed repeatedly - so I mean if something like this would lead to fewer lambda allocations/overhead:
```
drep = case Map.lookup poolId rePoolParams of
  Just d -> Map.lookup (raCredential (ppRewardAccount d)) reDelegatees
  Nothing -> Nothing
``` 
But don't know if it makes any difference at all, so feel free to ignore this.

lehins replied:
They are in fact identical, since this is exactly the implementation of `>>=` for `Maybe` and it will be optimized by the compiler. GHC is really good at this sort of stuff. For that reason I do prefer Daniel's implementation, since it is a bit more concise. Except, for further readability, I would also suggest removing redundant composition and then either splitting the line:
```suggestion
              Map.lookup poolId rePoolParams >>= \d -> 
                Map.lookup (raCredential (ppRewardAccount d)) reDelegatees
```
or switching to `do` syntax:
```suggestion
              d <- Map.lookup poolId rePoolParams
              Map.lookup (raCredential (ppRewardAccount d)) reDelegatees
```
But that's not critical by any means.

teodanciu reviewed (commented):
Looks good  to me, but will let @lehins approve since this is is performance-sensitive Umap stuff.

lehins reviewed (approved):
This looks great!
We'll definitely need to a few Imp tests for this and re-enable RATIFY conformance test once spec catches up. 
But, it is definitely OK to postpone that work until after this PR is merged.

Lucsanszky commented:
> This looks great! We'll definitely need to a few Imp tests for this and re-enable RATIFY conformance test once spec catches up. But, it is definitely OK to postpone that work until after this PR is merged.

Sounds good! I managed to tweak the conformance tests to filter for `Abstain` votes for SPOs, which is slightly better than disabling the `RATIFY` conformance tests altogether.
